### PR TITLE
DPDK Backend: Add support for switch statement in p4c-dpdk

### PR DESCRIPTION
**Pull request by usha1830**
This PR adds support for switch statement in p4c-dpdk

- This assumes that the EliminateSwitch pass is run and only expression in switch is of type tbl.apply().action.run
- Couple of tests started passing because of these changes. Adding reference outputs for these tests and updated the XFAIL list.
- Each case within switch is converted to **jmpa** instruction (jump on action run)

### CODE REVIEW

File: backends/dpdk/dpdk.def [hunk 1]
```diff
@@ -151,6 +151,29 @@ class DpdkJmpMissStatement : DpdkJmpStatement {
...
+}
+
+class DpdkJmpOnActionRunStatement : DpdkJmpActionStatement {
+#nodbprint
+#noconstructor
+    DpdkJmpOnActionRunStatement(cstring label, cstring act) :
+        DpdkJmpActionStatement("jmpa", label, act) { }
+}
+
+class DpdkJmpNotOnActionRunStatement : DpdkJmpActionStatement {
```
Comment (mbudiu-vmw) [inline backends/dpdk/dpdk.def:170]: that' an awkward name

File: backends/dpdk/dpdkArch.cpp [hunk 2]
```diff
@@ -1526,6 +1526,8 @@ const IR::Node* SplitP4TableCommon::postorder(IR::IfStatement* statement) {
 
 const IR::Node* SplitP4TableCommon::postorder(IR::SwitchStatement* statement) {
     auto expr = statement->expression;
+    if (!expr->is<IR::Member>())
```
Comment (mbudiu-vmw) [inline backends/dpdk/dpdkArch.cpp:1529]: could swap these or check if member is null below.
There is also a set of utilities in https://github.com/p4lang/p4c/blob/main/frontends/p4/tableApply.h that make it easier to detect this pattern.

File: backends/dpdk/dpdkHelpers.cpp [hunk 3]
```diff
@@ -16,6 +16,7 @@ limitations under the License.
 #include <iostream>
 #include "dpdkHelpers.h"
 #include "ir/ir.h"
+#include "frontends/p4/tableApply.h"
```
Comment (mbudiu-vmw) [inline backends/dpdk/dpdkHelpers.cpp:19]: I guess you know about it.

File: backends/dpdk/dpdkHelpers.cpp [hunk 4]
```diff
@@ -681,9 +682,52 @@ bool ConvertStatementToDpdk::preorder(const IR::MethodCallStatement *s) {
...
+        } else {
+            auto pe = caseLabel->label->to<IR::PathExpression>();
+            CHECK_NULL(pe);
+            label = Util::printf_format("label_action%d", next_label_id++);
+            add_instr(new IR::DpdkJmpOnActionRunStatement(label, pe->path->name.name));
+        }
+        labels.push_back(label);
+    }
+
+    if (!isdefaultPresent)
```
Comment (mbudiu-vmw) [inline backends/dpdk/dpdkHelpers.cpp:714]: don't we have a pass which ensures that this is always true (or that the cases are exhaustive?)

File: backends/dpdk/dpdkHelpers.cpp [hunk 5]
```diff
@@ -681,9 +682,52 @@ bool ConvertStatementToDpdk::preorder(const IR::MethodCallStatement *s) {
...
+    auto size = s->cases.size();
+    std::vector <cstring> labels;
+    cstring label;
+    bool isdefaultPresent = false;
+    cstring default_label = "";
+    auto end_label = Util::printf_format("label_endswitch%d", next_label_id++);
+    // Emit jmp on action run statements and collect labels for each case statement
+    for (unsigned i = 0; i < size; i++) {
+        auto caseLabel = s->cases.at(i);
+        if (caseLabel->label->is<IR::DefaultExpression>()) {
```
Comment (mbudiu-vmw) [inline backends/dpdk/dpdkHelpers.cpp:700]: we definitely have a pass that makes sure the default label is always last if it exists.

File: backends/dpdk/dpdkHelpers.cpp [hunk 6]
```diff
@@ -681,9 +682,52 @@ bool ConvertStatementToDpdk::preorder(const IR::MethodCallStatement *s) {
...
+            label = Util::printf_format("label_default%d", next_label_id++);
+            add_instr(new IR::DpdkJmpLabelStatement(label));
+            default_label = label;
+            isdefaultPresent = true;
+        } else {
+            auto pe = caseLabel->label->to<IR::PathExpression>();
+            CHECK_NULL(pe);
+            label = Util::printf_format("label_action%d", next_label_id++);
```
Comment (mbudiu-vmw) [inline backends/dpdk/dpdkHelpers.cpp:708]: are you sure these label names are unique?
could you use our existing tools to generate unique names (but then you have to do it for everything).

**Review by mbudiu-vmw — APPROVED** [review]

File: backends/dpdk/dpdkHelpers.cpp (hunk 4 above)
Comment (usha1830) [reply backends/dpdk/dpdkHelpers.cpp]: There is a frontend pass for adding default case. However, later the midend pass EliminateSwitch converts switch statements with other expressions to switch case with action_run expression. These newly created switch statements do not have a default case, the default case is also converted to an action and that action is marked as the const default action for the switch table.
https://github.com/p4lang/p4c/blob/main/midend/eliminateSwitch.h#L73

File: backends/dpdk/dpdkHelpers.cpp (hunk 5 above)
Comment (usha1830) [reply backends/dpdk/dpdkHelpers.cpp]: Yes. I updated the code to move the last case handling outside the loop.

File: backends/dpdk/dpdkHelpers.cpp (hunk 6 above)
Comment (usha1830) [reply backends/dpdk/dpdkHelpers.cpp]: Updated the code to use refmap->newName utility function to generate unique names.

**Comment by usha1830** [comment]
@mbudiu-vmw Thanks for the review! I have updated the code as per your comments. Please take a look.

File: backends/dpdk/dpdkHelpers.cpp (hunk 4 above)
Comment (mbudiu-vmw) [reply backends/dpdk/dpdkHelpers.cpp]: One solution is then to rerun this pass after the switch statements are eliminated...

**Review by mbudiu-vmw — APPROVED** [review]

**Pull request merged**